### PR TITLE
[IM] Convert call form IM to current ember ZCL implementation.

### LINKED ISSUE #5011: Enable IM on ZCL server (device) side.
#### Problem
We have enabled ZCL server on controller side, however, the device still missing ZCL support since we have to support various cluster and their commands, this is a high level issue that lists necessary steps.

 #### Proposed Solution

- [ ] Add a legacy mode to Message Layer (ExchangeManager) so it can handle both messages that is registered in ExchangeManager and unsupported message types are still handled by original message handlers.
- [ ] Generate the callback API translation by ZCL, this is a one-time work.
- [ ] Write the emberAfSendDefaultResponse and some functions like this for encode response in CHIP TLV format. And maintain the status in ZCL so all handlers can still working.

The detailed steps should be
1. Make lighting app working
2. Scale to all other apps / clusters.

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem
We need to enable IM, but we don't want to use a IM that only supports a small number of clusters & we don't want to drop the full ZCL stack for now. 

 #### Summary of Changes
This PR is focused on minimal support for ember ZCL library, that means a few number of clusters can work now, but not expected to be all.

This PR added

1. `src/app/zap-templates/templates/compatibility/compatibility-command-handlers.zapt` and other simiar files for "compatibility" between IM and ZCL. It will convert some callbacks from IM to the callback in ZCL and vice versa, there is also a list of clusters that need this step, it is based on current all-cluster-apps, that means, some of the clusters will be written in "native" CHIP style, and "reverse" compatibility will be added to current ember ZCL stack.
2. `src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt` for generate command dispatcher in IM.

After this PR, you can access most clusters via IM api and it works in most cases, some commands may not work and it will be fixed in the follow up PRs, this PR unblocks our next step on interaction model.

**There will be another PR for coexistence of current ZCL and the new IM.**

Tested with OnOff cluster commands and it works fine. The tests will be added in https://github.com/project-chip/connectedhomeip/issues/5005

 Fixes #5011

### CODE REVIEW

File: examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,2278 @@
...
+        return;
+    case ZCL_ON_OFF_CLUSTER_ID:
+        clusters::OnOff::DispatchServerCommand(apCommandObj, aCommandId, aEndPointId, aReader);
+        return;
+    case ZCL_SCENES_CLUSTER_ID:
+        clusters::Scenes::DispatchServerCommand(apCommandObj, aCommandId, aEndPointId, aReader);
+        return;
+    default:
+        // Unrecognized cluster ID, error status will apply.
+        // TODO: Encode response for Cluster not found
```
Comment (andy31415) [inline examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp:85]: A log message could be useful here

File: examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,2278 @@
...
+            while ((TLVError = dataTlv.Next()) == CHIP_NO_ERROR)
+            {
+                switch (TLV::TagNumFromTag(dataTlv.GetTag()))
+                {
+                case 0:
+                    TLVError = dataTlv.Get(percentOpen);
+                    break;
+                default:
+                    // Unsupported tag, ignore it.
+                    // TODO: Can be a log for INFO level
```
Comment (andy31415) [inline examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp:114]: All errors should be logged - I think we can do this instead of just TODO - the complexity does not seem high enough too warrant TODOs

File: examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.h [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+#include "call-command-handler.h"
+#include "callback.h"
+#include "cluster-id.h"
+#include "command-id.h"
+#include "util.h"
+
+#include <app/Command.h>
+#include <core/CHIPTLV.h>
+
+// The IM_HAVE_XXX_CLUSTER_YYY_COMMAND is used for compatibility handlers.
```
Comment (andy31415) [inline examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.h:32]: is there a TODO to remove them? They look odd inside namespaces (stutter and defines have no namespace scoping)

File: src/app/clusters/im-compatibility/compatibility-common.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
+/*
```
Comment (andy31415) [inline src/app/clusters/im-compatibility/compatibility-common.cpp:1]: could we find a more focused name than 'common'?

common/util/support/helpers are all names that are vague and are not clear when reading code why one would include 'common.h' or 'support.h'. 

Event the description of `Contains the common functions for compatibilty` does not seem to tell me what this is.

File: src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      // Just for compatibility, we will add array type support in IM later.
+      TLVError = dataTlv.GetDataPtr(const_cast<const uint8_t *&>({{asSymbol label}}));
+{{else}}
+      TLVError = dataTlv.Get({{asSymbol label}});
+{{/if}}
+      break;
+{{/zcl_command_arguments}}
+    default:
+      // Unsupported tag, ignore it.
+      // TODO: Can be a log for INFO level
```
Comment (andy31415) [inline src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt:29]: please add logs here and below.

**Review by andy31415 — COMMENTED** [review]
This is really neat!

File: examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.h (hunk 3 above)
Comment (erjiaqing) [reply examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.h:32]: Added a TODO

File: src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt (hunk 5 above)
Comment (erjiaqing) [reply src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt]: Yes, since this is just a single line of code.

File: src/app/clusters/im-compatibility/compatibility-common.cpp (hunk 4 above)
Comment (erjiaqing) [reply src/app/clusters/im-compatibility/compatibility-common.cpp]: Renamed it to ember-compatibility-functions to indicate it is used for maintain ember internal states.

File: src/app/clusters/im-compatibility/compatibility-command-handlers.cpp [hunk 6]
```diff
@@ -0,0 +1,841 @@
...
+namespace clusters {
+
+namespace BarrierControl {
+#if IM_HAVE_BARRIER_CONTROL_CLUSTER_BARRIER_CONTROL_GO_TO_PERCENT_COMMAND
+void OnBarrierControlGoToPercentCommandCallback(app::Command * command, EndpointId ZCLEndpointId, uint8_t percentOpen)
+{
+    Compatibility::SetupEmberAfObjects(command, ZCL_BARRIER_CONTROL_CLUSTER_ID, ZCL_BARRIER_CONTROL_GO_TO_PERCENT_COMMAND_ID,
+                                       ZCLEndpointId);
+    emberAfBarrierControlClusterBarrierControlGoToPercentCallback(percentOpen);
+    Compatibility::ResetEmberAfObjects();
```
Comment (vivien-apple) [inline src/app/clusters/im-compatibility/compatibility-command-handlers.cpp:50]: Why do you need `src/app/clusters/im-compatibility/compatibility-command-handlers.cpp` ? Can't you directly call `Compatibility::SetupEmberAfObjects(command, ZCL_BARRIER_CONTROL_CLUSTER_ID, commandId, endpointId);` at the beginning of `Clusters::BarrierControl::DispatchServerCommand`, `emberAf{{ClusterName}}Cluster{{CommandName}}Callback` in the switch/case, and `Compatibility::ResetEmberAfObjects();` at the end ?

File: src/app/clusters/im-compatibility/ember-compatibility-functions.cpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+EmberApsFrame imCompatibilityEmberApsFrame;
+EmberAfInterpanHeader imCompatibilityInterpanHeader;
+Command * currentCommandObject;
+} // namespace
+
+void SetupEmberAfObjects(Command * command, ClusterId clusterId, CommandId commandId, EndpointId endpointId)
+{
+    imCompatibilityEmberApsFrame.clusterId           = clusterId;
+    imCompatibilityEmberApsFrame.destinationEndpoint = endpointId;
+    imCompatibilityEmberApsFrame.sourceEndpoint      = endpointId;
```
Comment (vivien-apple) [inline src/app/clusters/im-compatibility/ember-compatibility-functions.cpp:45]: I think `sourceEndpoint` is basically `1` for now. You may want to use that instead of `endpointId`. For applications with multiple endpoints it will send the response on the wrong client endpoint otherwise.

Comment (vivien-apple) [inline src/app/clusters/im-compatibility/ember-compatibility-functions.cpp:45]: Can you also fill `imCompatibilityEmberApsFrame.sequence` (uint8_t) with some bits of the exchange id ? The sequence number is used for to map client callbacks to the sent command.

File: src/app/clusters/im-compatibility/ember-compatibility-functions.cpp [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    imCompatibilityEmberAfCluster.commandId          = commandId;
```
Comment (vivien-apple) [inline src/app/clusters/im-compatibility/ember-compatibility-functions.cpp:46]: Can you fill the `source` with the `nodeId` of the source here and similarly to `EmberApsFrame` can you fill `seqNum` with some bits of the exchange id ?

Comment (vivien-apple) [reply src/app/clusters/im-compatibility/ember-compatibility-functions.cpp:46]: In theory, we should also fill `mfgCode` and `mfgSpecific`. Not sure the IM has anything implemented for them yet ?

File: src/app/util/util.cpp [hunk 9]
```diff
@@ -770,6 +783,11 @@ EmberStatus emberAfSendImmediateDefaultResponseWithCallback(EmberAfStatus status
 
 EmberStatus emberAfSendImmediateDefaultResponse(EmberAfStatus status)
 {
+    if (chip::app::Compatibility::IMEmberAfSendImmediateDefaultResponseHandle(status))
+    {
+        // If the compatibility can handle this response
+        return EMBER_SUCCESS;
+    }
```
Comment (vivien-apple) [inline src/app/util/util.cpp:790]: Maybe you should move this code to https://github.com/project-chip/connectedhomeip/blob/8c93ec6b67f9c37cfeec9fa1d2cb93266eb4aa95/src/app/util/util.cpp#L776 since there could be multiple codepaths that leads to `emberAfSendImmediateDefaultResponseWithCallback` ?

File: src/app/clusters/im-compatibility/ember-compatibility-functions.cpp [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+                                                             imCompatibilityEmberAfCluster.commandId,
+                                                             (chip::app::Command::kCommandPathFlag_EndpointIdValid) };
+
+    chip::TLV::TLVType dummyType = chip::TLV::kTLVType_NotSpecified;
+    chip::TLV::TLVWriter writer  = currentCommandObject->CreateCommandDataElementTLVWriter();
+
+    SuccessOrExit(err = writer.StartContainer(chip::TLV::AnonymousTag, chip::TLV::kTLVType_Structure, dummyType));
+    SuccessOrExit(err = writer.EndContainer(dummyType));
+    SuccessOrExit(err = writer.Finalize());
+    SuccessOrExit(err = currentCommandObject->AddCommand(returnCommandParam));
```
Comment (vivien-apple) [inline src/app/clusters/im-compatibility/ember-compatibility-functions.cpp:76]: Can you point me to the code that handles the responses ? I suspect we may need a `DispatchClientCommand` method too to get the IM compatibility working fully for the default response.

File: src/app/zap-templates/templates/app/callback.zapt [hunk 11]
```diff
@@ -611,4 +611,3 @@ EmberStatus emberAfInterpanSendMessageCallback(EmberAfInterpanHeader * header, u
  *
  */
 bool emberAfStartMoveCallback();
-
```
Comment (vivien-apple) [inline src/app/zap-templates/templates/app/callback.zapt:614]: nit: Revert this change. That's likely just noise.

File: examples/lighting-app/lighting-common/BUILD.gn [hunk 12]
```diff
@@ -46,6 +48,14 @@ source_set("lighting-common") {
     "gen/callback-stub.cpp",
   ]
 
+  if (chip_app_use_interaction_model) {
+    sources += [
+      "${chip_root}/src/app/clusters/im-compatibility/compatibility-command-handlers.cpp",
+      "${chip_root}/src/app/clusters/im-compatibility/ember-compatibility-functions.cpp",
+      "gen/IMClusterCommandHandler.cpp",
+    ]
+  }
```
Comment (vivien-apple) [inline examples/lighting-app/lighting-common/BUILD.gn:57]: Can you add  a `else` case that adds `"gen/call-command-handler.cpp"`, `"${chip_root}/src/app/util/process-cluster-message.cpp"`, `"${chip_root}/src/app/util/process-global-message.cpp"` conditionally since this is basically what the current code is trying to replace. Or does it fails to link ?

**Review by vivien-apple — CHANGES_REQUESTED** [review]
Thanks for doing that! There is still some work but that's a great step forward.

File: examples/lighting-app/lighting-common/BUILD.gn (hunk 12 above)
Comment (erjiaqing) [reply examples/lighting-app/lighting-common/BUILD.gn:57]: These files are already in the general source list since we cannot handle the global messages now, the two stack should be able to co-exist for a while.

File: src/app/clusters/im-compatibility/compatibility-command-handlers.cpp (hunk 6 above)
Comment (erjiaqing) [reply src/app/clusters/im-compatibility/compatibility-command-handlers.cpp:50]: I'd like to make the Command object available for command handlers, and the future commands can be written in the IM style. Anyway, I think it is a good idea that we can call `Compatibility::SetupEmberAfObjects` and `Compatibility::ResetEmberAfObjects`  from the dispatcher.

File: src/app/clusters/im-compatibility/ember-compatibility-functions.cpp (hunk 7 above)
Comment (erjiaqing) [reply src/app/clusters/im-compatibility/ember-compatibility-functions.cpp]: Done

File: src/app/clusters/im-compatibility/ember-compatibility-functions.cpp (hunk 8 above)
Comment (erjiaqing) [reply src/app/clusters/im-compatibility/ember-compatibility-functions.cpp]: I did not found fields that is related to mfgCode in the CHIP spec, not sure if this is intentional.

File: src/app/util/util.cpp (hunk 9 above)
Comment (erjiaqing) [reply src/app/util/util.cpp]: Thanks!

**Comment by erjiaqing** [comment]
Just reordered the commits

```
git diff 99566d4 7f52839 
```

(no output)

File: src/app/clusters/im-compatibility/ember-compatibility-functions.cpp (hunk 8 above)
Comment (vivien-apple) [reply src/app/clusters/im-compatibility/ember-compatibility-functions.cpp]: > I did not found fields that is related to mfgCode in the CHIP spec, not sure if this is intentional or maybe I missed some words in the spec?

I have not found anything either in the spec.
One example of how it is used in the current code is to differentiate between commands. For example, the command id of `MfgSpecificPing` is `0x0000` which is the same than the command id of `ResetToFactoryDefaults`.
The `mfgCode` and `mfgSpecific` bits are then used to differentiate between those commands.
@bzbarsky-apple Are you aware of anything spec'ed for manufacturer specific extensions ?

File: src/app/clusters/im-compatibility/ember-compatibility-functions.cpp (hunk 10 above)
Comment (erjiaqing) [reply src/app/clusters/im-compatibility/ember-compatibility-functions.cpp]: Issue created: https://github.com/project-chip/connectedhomeip/issues/5096

File: src/app/clusters/im-compatibility/compatibility-command-handlers.cpp (hunk 6 above)
Comment (vivien-apple) [reply src/app/clusters/im-compatibility/compatibility-command-handlers.cpp]: Then it will be better to fix the current commands handlers to have a new parameter for it in a separate PR.
Please remove `src/app/clusters/im-compatibility`, it introduce a new generated folder in `src/` and creates a level of indirection that does not seems really needed.

Comment (erjiaqing) [reply src/app/clusters/im-compatibility/compatibility-command-handlers.cpp]: Issue: https://github.com/project-chip/connectedhomeip/issues/5098

**Review by vivien-apple — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Comment by erjiaqing** [comment]
@andy31415 @mspang @bzbarsky-apple @woody-apple PTAL, thanks

**Comment by bzbarsky-apple** [comment]
I'll try to look tomorrow (Wednesday), but an important question: how much do we care about correctness here?  The code is not verifying that it's getting the right types, for example (e.g. uses TLVReader's `Get` for integers without checking the size of the integer, which has some ... let's charitably say "interesting" ... behavior).

**Comment by erjiaqing** [comment]
> I'll try to look tomorrow (Wednesday), but an important question: how much do we care about correctness here?  The code is not verifying that it's getting the right types, for example (e.g. uses TLVReader's `Get` for integers without checking the size of the integer, which has some ... let's charitably say "interesting" ... behavior).

The reader class should take care of this. And we only need to care about the error it returns( "TLVError" in this patch).
I think it is ok if we ask to read a int64 value but only gives a int32 variable when the value is in int32 range actually?

**Review by andy31415 — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
> The reader class should take care of this.

It currently does not.

> I think it is ok if we ask to read a int64 value but only gives a int32 variable when the value is in int32 range actually?

Sure, but the problems start when we ask to read an int8 value but the TLV contains a value that does not actually fit in 8 bits.  That read will succeed, which is broken.  You're right that we should solve this in the TLV reader in general, but no one has had time to do that yet.

**Comment by erjiaqing** [comment]
> > The reader class should take care of this.
> 
> It currently does not.
> 
> > I think it is ok if we ask to read a int64 value but only gives a int32 variable when the value is in int32 range actually?
> 
> Sure, but the problems start when we ask to read an int8 value but the TLV contains a value that does not actually fit in 8 bits. That read will succeed, which is broken. You're right that we should solve this in the TLV reader in general, but no one has had time to do that yet.



> > The reader class should take care of this.
> 
> It currently does not.
> 
> > I think it is ok if we ask to read a int64 value but only gives a int32 variable when the value is in int32 range actually?
> 
> Sure, but the problems start when we ask to read an int8 value but the TLV contains a value that does not actually fit in 8 bits. That read will succeed, which is broken. You're right that we should solve this in the TLV reader in general, but no one has had time to do that yet.

Checked that TLV does not expose such api to get size info of the value.

Just created an issue for this: https://github.com/project-chip/connectedhomeip/issues/5152

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@erjiaqing Can you take a peek at the conflict please?

**Comment by bzbarsky-apple** [comment]
> Checked that TLV does not expose such api to get size info of the value.

Right, the only safe thing to do right now is to get into a 64-bit int and check the value yourself.  And even then signed vs unsigned will trip you up.

> Just created an issue for this: #5152

Alright.  Is the plan to resolve that first, or to land this PR as is and then follow up in that issue?

**Review by mspang — APPROVED** [review]

**Review by mspang — CHANGES_REQUESTED** [review]
Needs a ZAP generate.

**Review by mspang — APPROVED** [review]

**Comment by erjiaqing** [comment]
> Needs a ZAP generate.

Done
